test/zdtm/static/file_cloexec.c

```diff
@@ -0,0 +1,63 @@
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <sys/vfs.h>
+#include <unistd.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc	= "Check FD_CLOEXEC flag";
+const char *test_author	= "Nicolas Viennot <Nicolas.Viennot@twosigma.com>";
+
+#define err(exitcode, msg, ...) ({ pr_perror(msg, ##__VA_ARGS__); exit(exitcode); })
+
+static void assert_fd_flags(int fd, int mask, int value)
+{
+	int flags = fcntl(fd, F_GETFD);
+	if (flags == -1)
+		err(1, "Can't get fd flags");
+
+	if ((flags & mask) != value) {
+		fail("fd flags mismatch");
+		exit(1);
+	}
+}
+
+int main(int argc, char *argv[])
+{
+	int fd1, fd2, fd3, fd4;
+
+	test_init(argc, argv);
+
+	fd1 = open("/", O_RDONLY | O_CLOEXEC);
+	if (fd1 < 0)
+		err(1, "Can't open()");
```

Comment:
pls use pr_perror() to log errors

Reply:
There's a define on top that invokes pr_perror() and then exit().
Is it acceptable?

Reply:
@nviennot I don't like this, but I think this is acceptable. I understand the problem what you are trying to solve here. What do you think about the following more generic solution:

```
diff --git a/test/zdtm/lib/zdtmtst.h b/test/zdtm/lib/zdtmtst.h
index 1fbf795bf..ffe18469c 100644
--- a/test/zdtm/lib/zdtmtst.h
+++ b/test/zdtm/lib/zdtmtst.h
@@ -122,11 +122,17 @@ extern int write_pidfile(int pid);
 /* message helpers */
 extern int test_log_init(const char *outfile, const char *suffix);
 extern int zdtm_seccomp;
-#define pr_err(format, arg...) \
-       test_msg("ERR: %s:%d: " format, __FILE__, __LINE__, ## arg)
-#define pr_perror(format, arg...)      \
+#define pr_err(format, arg...) ({ \
+       test_msg("ERR: %s:%d: " format, __FILE__, __LINE__, ## arg); \
+       1;  \
+})
+
+#define pr_perror(format, arg...)      ({ \
        test_msg("ERR: %s:%d: " format " (errno = %d (%s))\n", \
-               __FILE__, __LINE__, ## arg, errno, strerror(errno))
+               __FILE__, __LINE__, ## arg, errno, strerror(errno)); \
+       1; \
+})
+
 #define fail(format, arg...)   \
        test_msg("FAIL: %s:%d: " format " (errno = %d (%s))\n", \
                 __FILE__, __LINE__, ## arg, errno, strerror(errno))
diff --git a/test/zdtm/static/env00.c b/test/zdtm/static/env00.c
index 1feabfa9f..cb3b3417f 100644
--- a/test/zdtm/static/env00.c
+++ b/test/zdtm/static/env00.c
@@ -16,10 +16,8 @@ int main(int argc, char **argv)
 
        test_init(argc, argv);
 
-       if (setenv(envname, test_author, 1)) {
-               pr_perror("Can't set env var \"%s\" to \"%s\"", envname, test_author);
-               exit(1);
-       }
+       if (setenv(envname, test_author, 1))
+               return pr_perror("Can't set env var \"%s\" to \"%s\"", envname, test_author);
 
        test_daemon();
        test_waitsig();
```

Reply:
I think we should keep the `err()` semantics: http://man7.org/linux/man-pages/man3/err.3.html for the following reasons:
1) `return 1` works only when called from `main()`. When called from another function, `return -1` makes more sense. `exit(1)` always works.
2) We should try to keep functions that have well known semantics to ease readability.